platform/mbed_retarget.h

```diff
@@ -66,29 +73,63 @@ namespace mbed {
 
 class FileHandle;
 class DirHandle;
-std::FILE *mbed_fdopen(FileHandle *fh, const char *mode);
+
+/** Bind an mbed FileHandle to a POSIX file descriptor */
+int bind_to_fd(mbed::FileHandle *fh);
+
+/** Unbind an mbed FileHandle from a POSIX file descriptor
+ * Intended for internal use - normally unbinding happens
+ * as part of close() or fclose().
+ */
+int unbind_from_fd(int fd, mbed::FileHandle *fh);
+
+/** Targets may implement this to change stdin, stdout, stderr.
+ *
+ * If the application hasn't provided mbed_override_console, this is called
+ * to give the target a chance to specify a FileHandle for the console.
+ *
+ * If this is not provided or returns NULL, the console will be:
+ *   - UARTSerial if configuration option "platform.stdio-buffered-serial" is
+ *     true and the target has DEVICE_SERIAL;
+ *   - Raw HAL serial via serial_getc and serial_putc if
+ *     "platform.stdio-buffered-serial" is false and the target has DEVICE_SERIAL;
+ *   - stdout/stderr will be a sink and stdin will input a stream of 0s if the
+ *     target does not have DEVICE_SERIAL.
+ *
+ * @param fd file descriptor - STDIN_FILENO, STDOUT_FILENO or STDERR_FILENO
+ * @return  pointer to FileHandle to override normal stream otherwise NULL
+ */
+FileHandle* mbed_target_override_console(int fd);
```

Comment:
a use case for a target overriding default console? isn't this app specific, and we provide default implementation?

Reply:
The use case is when the board knows that it's not appropriate for trace to go to a UART.

#5726 contains an example of that - there's one UART on the SoC which is connected to an onboard modem, so there's no UART spare for trace. If the board can't provide an alternative console, it would have to provide a fake serial port. And ensure all apps were modified to not use stdout.

Compare #5356, which is currently a support library/application thing that sends the console to Segger RTT. But conceivably a particular board might want that Segger RTT to be the default, like #5726.

Regardless of what the board does (use default serial, or something else), the application can always make the final choice.

Reply:
Not to bikeshed but the api is a bit verbose. Suggestion: retarget_console()? Also, the config for UARTSerial seems like an optimization that doesn't need to be part of this patch (not that its bad).